### ISSUE TITLE
Chocolate Doom should have an Emscripten port

### ISSUE THREAD

**Issue by fragglet** [chocolate-doom/chocolate-doom#538]
First of all, I played around with Emscripten several weeks ago and although I had success getting other things to work, I didn't manage to get far with Chocolate Doom. That said, I do believe that an Emscripten (ie. Javascript) port is possible and we ought to support it.

Brain dump of current thoughts:

* One big problem is restructuring the code to be event-based. With an Emscripten program you cannot have infinite loops, which crash the browser. In practice this means restructuring d_loop.c and txt_desktop.c so that their event loops are event driven by the Javascript event loop.

* Emscripten has built-in support for the SDL API, but this is really a half-baked reimplementation of SDL rather than a port of the code. A whole bunch of the API functions are missing. I got as far as getting one of the libtextscreen demos to run, but ultimately just hit a blank screen, because it didn't support palettized surfaces.

* However, [the SDL developers have done a proper port of SDL2](https://twitter.com/icculus/status/550022999160913920) to run on Emscripten. This almost certainly offers a better basis for a port, and means that any Emscripten port is going to be based on sdl2-branch (#450). The Emscripten port isn't in the current release (2.0.3) so we'd have to use a version of SDL2 from Mercurial HEAD.

**Comment by fragglet** [comment]
Also: philosophically, this fits within Chocolate Doom's goal for maximum portability. There are now operating systems like Chrome OS and Firefox OS that offer nothing but a web browser. Emscripten offers potentially the only way to reach these platforms. But the concept of running the game entirely in a browser is also just really nice: every OS has a web browser and can run Javascript.

Ideally I'd like to see a completed port made available as a download (.html and .js files, along with the usual docs), as well as available as a page on the website. History has shown that we probably won't be able to put it up with the shareware IWADs, but maybe we can provide FreeDM as an alternative. Providing a Chrome app would also be a nice touch.

**Comment by hifi** [comment]
3.0.0 builds and runs on Emscripten with minor tweaks without sound as far as my testing goes. I'll see if I can clean it up to the point of being mergeable.

SDL_mixer is not yet enabled from Emscripten Ports so that needs some extra work for now.

**Comment by hifi** [comment]
Okay so this commit adds enough to compile and run `chocolate-doom`: https://github.com/hifi/chocolate-doom/commit/9b3ea89cedcbb2bf590b1e0b2f7b2ae26bd82ea3

Setup changes were not finished. And for the shareware IWAD redistribution, I may have solved it by using `doom19s.zip` directly as the data source rather than distributing an extracted version.

Threw together a GitHub Page as a demo, seems to work on Chrome and Firefox except the audio has some issues: https://hifi.github.io/playdoom/

**Comment by Zuccace** [comment]
I can confirm. The above works with Firefox 61 on Linux. Audio seems to have some delay sometimes. Maybe Firefox opens and closes sockets to pulseaudio with every sound effect played?

**Comment by fragglet** [comment]
Fantastic news, thank you! I was expecting that to be far more work.

A couple of points of feedback:
* Ideal way of doing this would be if you could extend/refactor `d_loop.c` in a way that hides the underlying Emscripten details. That file deals with the game loop and it should be straightforward to add something like a `D_GameLoop()` function that either calls `emscripten_set_main_loop` or just uses a normal `while` loop in the non-Emscripten case. As much as possible I'd like to avoid littering the source code with lots of conditional #ifdefs.
* Great to see you've put up a playable demo, but I'd suggest that instead of `doom19s.zip` you instead put up a copy of [FreeDM](https://freedoom.github.io/). Past experience has been that id/Bethesda do not take kindly to Doom-in-a-browser. I know it's stupid but that seems to be their position. My plan was that it should be possible for the user to "upload" an IWAD into the browser's local storage.

Looking forward to PR. Thanks!

**Comment by AXDOOMER** [comment]
Works for me on Waterfox 56.2.2 on Debian. The audio issue is different. There's no audio delay, but the sounds are cut off (for example: when I shoot the shotgun, I hear it fire but the sound cuts off before I can hear the cocking). Also, the melt screen is not displayed.

**Comment by linguica** [comment]
I disagree about the Doom shareware thing. Let's remember https://doomwiki.org/wiki/Doom_on_the_Web and the specific reason Zenimax sent a takedown notice, namely, that it was distributing copyrighted assets. Since it's now just taking an existing (and presumably totally legal) doom19s.zip and extracting the IWAD, it would be good to know right off the bat if Zenimax is going to make the same legal argument.

**Comment by fragglet** [comment]
@linguica I'd rather just not have the debate. Fact is, it's clear they don't want Doom playable on a webpage and I doubt the particular technical details of how it's done are going to make much of a difference to them. Rather than potentially piss them off or risk any kind of takedown notice I'd rather just put up something unambiguously okay, like FreeDM. But it's @hifi who's hosting the page so it's up to him.

Another bug report: The screen scaling doesn't appear to be working right yet. If you look at the weapon while it's bobbing you'll see a rippling pattern. I suspect the linear scaling part of the screen scaling process is being skipped.

**Comment by linguica** [comment]
For posterity, the v1.8 Doom shareware included a LICENSE.DOC that had the following:

>3   Electronic Distribution is Permitted: ID grants to you the right to distribute, royalty free 
>and by electronic means only, the Software; provided, however the Software must be so 
>distributed only in compressed format.

**Comment by hifi** [comment]
For hosting on chocolate-doom.org, FreeDM is probably the way to go as it would also promote a free software data alternative.

I'll look into a more clean refactor.

**Comment by hifi** [comment]
Updated the branch with working melt and refactored the Emscripten ifdefs into `d_loop.c` as requested (also updated the demo). Looking into making game exit not burn the CPU and fixing chocolate-doom-setup, then making local storage working for configuration changes so that it would effectively just work.

Still don't know what's wrong with sound effects being cut off.

**Comment by hifi** [comment]
You can do pretty weird stuff with a browser these days.

https://www.youtube.com/watch?v=epAPVWU7bH8

Also the in the demo, if you quit the game you end up in setup and vice versa so you can actually configure things. Had to put huge on screen buttons to configure the controller on my phone from within setup.

**Comment by turol** [comment]
Someone at Cloudflare did the hard parts already: https://github.com/cloudflare/doom-wasm
There's also a [blog post](https://blog.cloudflare.com/doom-multiplayer-workers/). Now someone just needs to integrate it back. I think someone already volunteered on IRC.

As a further enhancement there's [HumbleNet](https://github.com/HumbleNet/HumbleNet) for better networking in browser.

**Comment by turol** [comment]
That repo has removed their history but it looks like they forked around 74773f9a5bc7cbeb3e7807ce28229a7644416cd2

**Comment by midzer** [comment]
Hi all,
on this weekend I got an Emscripten build to work.

Main issue was music support. Apparently, SDL Mixer can't handle the `mus` format inside `.wad` files. After digging around things finally worked by enabling Native MIDI support in DOOM config using Timidity software emulation.

Anyway, here's my fork with all changes: https://github.com/midzer/chocolate-doom/tree/emscripten

Doom 1 Shareware online: https://midzer.de/wasm/doom/

Have a great time,
midzer

**Comment by turol** [comment]
@midzer Do you intend to turn this into a PR? Some of those changes look like they might break other platforms.

**Comment by midzer** [comment]
It would be nice to contribute something back. Whether some1 picks things from my changes or I can try a PR.

**Comment by turol** [comment]
@jbaicoianu #1717 implemented some Emscripten stuff. What is missing? A nice launcher? A web-hosted demo? Some kind of multiplayer?

**Comment by jbaicoianu** [comment]
I'm using the WASM build for a larger project which lets you browse, view, and play user-created Doom wads in the browser, I'll be releasing that sometime in the next few weeks, but that's a pretty complicated demo so probably doesn't fit what you're looking for here.

It might be nice to have a version on the Chocolate Doom website which lets you play Doom shareware or Freedom. I've got a simpler demo which lets you pick from those, or you can drag and drop your own IWAD files to play those if you have a locally installed commercial version.

If that's of interest, I can clean it up and contribute it - is the chocolate Doom website in a GitHub repo somewhere as well?
